zkvm/docs/zkvm-spec.md

```diff
@@ -1906,23 +1968,22 @@ ZkVM ensures transaction uniqueness this way:
 * `issue` does not consume zero value
 * `finalize` does not consume zero value
 * `claim/borrow` can produce an arbitrary value and its negative at any point
-* Each UTXO ID is defined as `Hash(contract, txid)`, that is contents of the contract are not unique, but the new UTXO ID is defined by transaction ID, not vice versa.
+* Each UTXO ID is defined as `Hash(contract)`, that is contents of the contract are unique due to a contract’s anchor.
 * Transaction ID is a hash of the finalized log.
-* When VM finishes, it checks that the log contains either an [input](#input-entry) or a [nonce](#nonce-entry), setting the `uniqueness` flag.
-* Outputs are encoded in the log as snapshots of contents. Blockchain state update hashes these with transaction ID when generating UTXO IDs.
-* Inputs are encoded in the log as their UTXO IDs, so the blockchain processor knows which ones to find and remove.
+* Newly created contracts consume the "last anchor" tracked by VM and set it to their contract ID.
+* When VM finishes, it checks that "last anchor" is present, which means that the log contains either an [input](#input-entry) or a [nonce](#nonce-entry).
+* Inputs and outputs are represented in the log as contract IDs which the blockchain state machine should remove/insert accordingly.
 
 **Pros:**
 
-Huge pro: recipient can know upfront and provide full spec for the _compressed_ contract+value specification, so it can be revealed behind a shuffle.
-
-Handling values becomes much simpler — they are just values. So we can issue, finalize and even “claim” a value in a straightforward manner.
-
 The values are simply pedersen commitments (Q,A) (quantity, flavor), without any extra payload.
+This makes handling them much simpler: we can issue, finalize and even “claim” a value in a straightforward manner.
 
 **Con:**
 
-UTXO IDs are not known until the full transaction log is formed. This could be not a big deal, as we cannot really plan for the next transaction until this one is fully formed and published. Also, in a joint proof scenario, it’s even less reliable to plan the next payment until the MPC is completed, so requirement to wait till transaction ID is determined may not be a big deal.
+UTXO IDs are not generally known without fuller view on transaction flow. This could be not a big deal, as we cannot really plan for the next transaction until this one is fully formed and published. Also, in a joint proof scenario, it’s even less reliable to plan the next payment until the MPC is completed, so requirement to wait till transaction ID is determined may not be a big deal.
```

Comment:
Expand MPC?

Reply:
```suggestion
UTXO IDs are not generally known without fuller view on transaction flow. This could be not a big deal, as we cannot really plan for the next transaction until this one is fully formed and published. Also, in a joint proof scenario, it’s even less reliable to plan the next payment until the multi-party protocol is completed, therefore waiting for transaction ID to be determined becomes a requirement.
```

Reply:
:+1: